### DISCUSSION TITLE
Persist Middleware | Wait for store to be hydrated but only return once

### DISCUSSION THREAD

**Discussion by agustingabiola** [pmndrs/zustand#2531] [Q&A] [answered]
Hello community. I'm struggling to avoid certain re-renders in my application using the persist middleware. One of the things I noticed is that after the store is hydrated from localstorage a new render takes place. In my particular case I only need the "initial hydrated" values only once to initialize a form and even though the values I request change during the wizard I only need them for the initial load.

```
// store.ts

export const wizardStore =
  create<T>()(
    persist(
      (set, get) => ({
        user: { name: "", ....},
      {
        name: "key"
      }
    )
  );
```

```
// component.tsx

export default Component() => {

  ...

  const userValues = wizardStore(
    (state) => state.user,
   () => wizardStore.persist.hydrated()
  );
  
  ...
```

In this case `wizardStore.persist.hydrated()` is true when it's evaluated after hydration (makes sense), so I tried creating a ref with the objective that it tries another time after being hydrated but that didn't update the value even though I checked old and new states and the information of the user name is there.

```
// componentWithRef.tsx

export default Component() => {

  ...

  const userValues = wizardStore(
    (state) => state.user,
    () => {
        if (wizardStore.persist.hydrated() && !ref.current) {
           ref.current = true;
           return false;
       }
       return true;
    }
  );
  
  ...
```

Adding this amount of code just to avoid a couple of re-renders seems overkill  but I wanted to try and understand why it's not updated only once after store hydrated.

Thanks in advance.

Stack:

NextJS 14 Zustand 4.5.2

**Comment by dbritto-dev** [comment, marked as answer]
@agustingabiola  here you go -> https://stackblitz.com/edit/nextjs-and-zustand-persist?file=README.md btw, there's no way to avoid the loading state

**Reply by agustingabiola** [reply]
@dbritto-dev Thank you for the swift response. I was not expecting to have to use react context but from what I read is unavoidable. Thanks again!
